File: ThreadLab1Java/src/ThreadManager.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class ThreadManager implements Runnable {
    private final short threadCount;
    private Scanner scanner;

    public ThreadManager(short threadCount) {
        this.threadCount = threadCount;
        this.scanner = new Scanner(System.in);
    }

    @Override
    public void run() {
        List<Integer> durations = getDurations(threadCount);

        List<Integer> steps = getSteps(threadCount);

        List<MainThread> mainThreads = new ArrayList<>();

        for (int i = 0; i < threadCount; i++) {
            MainThread mainThread = new MainThread(steps.get(i), durations.get(i));
            mainThreads.add(mainThread);

            Thread thread = new Thread(mainThread);
            thread.start();
        }

        BreakThread breakThread = new BreakThread(mainThreads);
        Thread stopperThread = new Thread(breakThread);

        stopperThread.start();

    }

    private List<Integer> getDurations(short threadCount) {
        while (true) {
            System.out.printf("Enter runtime in milliseconds for %d thread(s): %n", threadCount);
            String input = scanner.nextLine().trim();

            if (input.isEmpty()) continue;

            String[] stringDurations = input.split("\\s+");

            if (stringDurations.length != threadCount) {
                System.out.printf("Error: please enter %d numbers.%n", threadCount);
                continue;
            }

            List<Integer> durations = new ArrayList<>();
            boolean allValid = true;

            for (String stringDuration : stringDurations) {
                try {
                    int duration = Integer.parseInt(stringDuration);
                    if (duration > 0) {
                        durations.add(duration);
                    } else {
                        System.out.printf("Error: '%s' is not a valid integer.%n", stringDuration);
                        allValid = false;
                        break;
                    }
                } catch (NumberFormatException e) {
                    System.out.printf("Error: '%s' is not a valid integer.%n", stringDuration);
                    allValid = false;
                    break;
                }
            }

            if (allValid) {
                return durations;
            }
        }
    }

    private List<Integer> getSteps(short threadCount) {
        while (true) {
            System.out.printf("Enter the step for %d thread(s): %n", threadCount);
            Scanner scanner = new Scanner(System.in);
            String input = scanner.nextLine().trim();

            if (input.isEmpty()) continue;

            String[] parts = input.split("\\s+");

            if (parts.length != threadCount) {
                System.out.printf("Error: please enter %d numbers.%n", threadCount);
                continue;
            }

            List<Integer> steps = new ArrayList<>();
            boolean allValid = true;

            for (String part : parts) {
                try {
                    int step = Integer.parseInt(part);
                    steps.add(step);
                } catch (NumberFormatException e) {
                    System.out.printf("Error: '%s' is not a valid integer.%n", part);
                    allValid = false;
                    break;
                }
            }

            if (allValid) {
                return steps;
            }
        }
    }
}

File: ThreadLab1C#/ThreadLab1/ThreadManager.cs
using System.Diagnostics;
using System.Threading;

namespace ThreadLab1
{
    internal class ThreadManager
    {
        private readonly short threadCount;

        public ThreadManager(short threadCount)
        {
            this.threadCount = threadCount;
        }

        public void Start()
        {
            List<int> durations = GetThreadDurations(threadCount);

            List<int> steps = GetSteps(threadCount);

            List<MainThread> mainThreads = new List<MainThread>();

            for (int i = 0; i < threadCount; i++)
            {
                var mainThread = new MainThread(steps[i], durations[i]);
                mainThreads.Add(mainThread);

                Thread thread = new Thread(mainThread.Run);
                thread.Start();
            }

            BreakThread breakThread = new BreakThread(mainThreads);
            Thread stopperThread = new Thread(breakThread.Run);

            stopperThread.Start();
        }

        private static List<int> GetThreadDurations(short threadCount)
        {
            while (true)
            {
                Console.WriteLine($"Введіть час роботи в мілісекундах для {threadCount} потоків: ");
                string input = Console.ReadLine() ?? string.Empty;

                string[] strDurations = input.Split(' ', StringSplitOptions.RemoveEmptyEntries);

                if (strDurations.Length != threadCount)
                {
                    Console.WriteLine($"Помилка: потрібно ввести {threadCount} чисел.");
                    continue;
                }


                List<int> durations = new List<int>();
                bool allValid = true;

                foreach (string strDuration in strDurations)
                {
                    if (int.TryParse(strDuration, out int duration) && duration > 0)
                    {
                        durations.Add(duration);
                    }
                    else
                    {
                        Console.WriteLine($"Помилка: '{strDuration}' не є валідним цілим числом.");
                        allValid = false;
                        break;
                    }
                }

                if (allValid)
                {
                    return durations;
                }
            }
        }

        private static List<int> GetSteps(int threadCount)
        {
            while (true)
            {
                Console.WriteLine("Введіть кроки для обчислення: ");
                string input = Console.ReadLine() ?? string.Empty;

                string[] parts = input.Split(' ', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length != threadCount)
                {
                    Console.WriteLine($"Помилка: потрібно ввести {threadCount} чисел.");
                    continue;
                }

                List<int> steps = new List<int>();
                bool allValid = true;

                foreach (string part in parts)
                {
                    if (int.TryParse(part, out int step))
                    {
                        steps.Add(step);
                    }
                    else
                    {
                        Console.WriteLine($"Помилка: '{part}' не є валідним цілим числом.");
                        allValid = false;
                        break;
                    }
                }

                if (allValid)
                {
                    return steps;
                }
            }
        }

    }
}
